crates/llama-cpp-server/src/supervisor.rs

```diff
@@ -87,30 +92,64 @@ impl LlamaCppSupervisor {
+                let mut stderr =
+                    BufReader::new(process.stderr.take().expect("Failed to get stderr")).lines();
+                let mut error_lines = VecDeque::with_capacity(100);
+
+                let wait_handle = process.wait();
+
+                while let Ok(Some(line)) = stderr.next_line().await {
+                    if !line.contains("GET /health") {
+                        if error_lines.len() >= 100 {
+                            error_lines.pop_front();
+                        }
+                        error_lines.push_back(line);
+                    }
+                }
+
+                let status_code = wait_handle.await.ok().and_then(|s| s.code()).unwrap_or(-1);
 
                 if status_code != 0 {
                     warn!(
                         "llama-server <{}> exited with status code {}, args: `{}`",
                         name, status_code, command_args
                     );
-                    let mut stderr = process
-                        .stderr
-                        .take()
-                        .map(tokio::io::BufReader::new)
-                        .map(|reader| reader.lines())
-                        .expect("Failed to read stderr");
-
-                    while let Ok(Some(line)) = stderr.next_line().await {
-                        warn!("<{}>: {}", name, line);
-                    }
 
-                    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                    let error_message = error_lines.into_iter().fold(
+                        String::from("Recent llama-cpp errors:\n"),
+                        |mut acc, line| {
+                            acc.push_str(&format!("{}\n", line));
+                            acc
+                        },
+                    );
+
+                    match status_code {
+                        0 => (),
+                        1 | -1 => {
+                            eprintln!("{}", error_message);
+                            if let Some(solution) = analyze_error_message(&error_message) {
+                                let solution_lines: Vec<_> = solution.split('\n').collect();
+                                let msg = tabby_common::terminal::InfoMessage::new(
+                                    "ERROR",
+                                    tabby_common::terminal::HeaderFormat::BoldRed,
+                                    &solution_lines,
+                                );
+                                msg.print();
+                            }
+                            eprintln!(
+                                "llama-server <{}> encountered a fatal error. Exiting service. Please check the above logs and suggested solutions for details.",
+                                name
+                            );
+                            std::process::exit(1);
+                        }
+                        _ => {
+                            warn!(
```

Comment:
how about printing the first or some of the retried error messages, so we can gather unknown error messages when users experience them or if they are caught in a retry loop.

WDYT @wsxiaoys

Reply:
SGTM